fe/fe-core/src/main/java/org/apache/doris/qe/Coordinator.java

```diff
@@ -1850,9 +1850,19 @@ private void getExecHostPortForFragmentIDAndBucketSeq(TScanRangeLocations seqLoc
         private void computeScanRangeAssignmentByBucket(
                 final OlapScanNode scanNode, ImmutableMap<Long, Backend> idToBackend, Map<TNetworkAddress, Long> addressToBackendID) throws Exception {
             if (!fragmentIdToSeqToAddressMap.containsKey(scanNode.getFragmentId())) {
-                // The bucket shuffle join only hit when the partition is one. so the totalTabletsNum is all tablet of
-                // one hit partition. can be the right bucket num in bucket shuffle join
-                fragmentIdToBucketNumMap.put(scanNode.getFragmentId(), (int) scanNode.getTotalTabletsNum());
+                // In bucket shuffle join, we have 2 situation.
+                // 1. Only one partition: in this case, we use scanNode.getTotalTabletsNum() to get the right bucket num
+                //    because when table turn on dynamic partition, the bucket number in default distribution info
+                //    is not correct.
+                // 2. Table is colocated: in this case, table could have more than one partition, but all partition's
+                //    bucket number must be same, so we use default bucket num is ok.
+                int bucketNum = 0;
+                if (scanNode.getOlapTable().isColocateTable()) {
```

Comment:
What if the colocate group is unstable?

Reply:
already handled in [DistributedPlanner ](https://github.com/apache/incubator-doris/blob/e701c057dc675babe7a859e151287fea7cd5c850/fe/fe-core/src/main/java/org/apache/doris/planner/DistributedPlanner.java#L618)